hysr/extra_balls.py

```diff
@@ -0,0 +1,268 @@
+        """
+        Requests the corresponding pam_mujoco instance to burst
+        """
+        self._handle.burst(nb_iterations)
+
+    def get_contacts(
+        self, index: ListOrIndex = None
+    ) -> typing.Sequence[context.ContactInformation]:
+        """
+        Returns the list contact informations between the balls and the contact
+        object (see argument "contact" of the constructor), i.e. an object with
+        attributes:
+        - contact occured : if true, at least one contact has occured
+        - position: if contact occured, the 3d position of the first contact
+        - time_stamp: if contact occured, the time stamp of the fist contact
+        - minimal_distance: if contact did not occure, the minimal distance
+                            between the two items
+        - disabled : true if contact detection has been disabled
+        Note that once a contact occured, the ball is no longer controlled by
+        o80 (i.e. the load method of this class will have no effect for the
+        corresponding ball), but by mujoco engine (until the method
+        reset_contacts of this class is called)
+        """
+        return list(map(self._handle.get_contact, self._get_segment_ids(index)))
+
+    def reset_contacts(self, index: ListOrIndex = None) -> None:
+        """
+        Reset the contacts, i.e. get_contacts will return instances that
+        indicates no contact occured. Also, restore o80 control of the ball.
+        """
+        list(map(self._handle.reset_contact, self._get_segment_ids(index)))
+
+    def activate_contacts(self, index: ListOrIndex = None) -> None:
+        """
+        Contacts will not be ignored.
+        """
+        list(map(self._handle.activate_contact, self._get_segment_ids(index)))
+
+    def deactivate_contacts(self, index: ListOrIndex = None) -> None:
+        """
+        Contacts will be ignored
+        """
+        list(map(self._handle.deactivate_contact, self._get_segment_ids(index)))
+
+    def set_trajectory_getter(self, trajectory_getter: TrajectoryGetter) -> None:
+        """
+        Overwrite the current instance of trajectory_getter
+        (that will be used by the method load_trajectories).
+        """
+        self._trajectory_getter = trajectory_getter
+
+    def get(self) -> ExtraBallsState:
+        """
+        Returns the current state of this extra balls set
+        """
+        observation = self._frontend.latest()
+        racket_cartesian = observation.get_extended_state().robot_position
+        contacts = observation.get_extended_state().contacts
+        iteration = observation.get_iteration()
+        time_stamp = observation.get_time_stamp()
+        state = observation.get_observed_states()
+        balls = [state.get(index) for index in range(self._size)]
+        positions = []
+        velocities = []
+        [
+            (positions.append(b.get_position()), velocities.append(b.get_velocity()))
+            for b in balls
+        ]
```

Comment:
I think it would be better to use a normal for loop here, so it's easier to follow what the code is doing.